### DISCUSSION TITLE
How to limit the CPU for docker-based workspace

### DISCUSSION THREAD

**Discussion by jizusun** [coder/coder#9012] [General] [answered]
Recently we're provider our docker-based Coder for more than 15+ developers with a 32 core VM, but sometimes the VM becomes unresponsive ( VSCode disconnected, even SSH into the VM is not possible).

With the following 
```hcl
resource "docker_image" "main" {
  name = "coder-${data.coder_workspace.me.id}"
  build {
    context = "./build"
    # https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/image
    # https://docs.docker.com/config/containers/resource_constraints/#configure-the-default-cfs-scheduler
    cpu_quota  = 150000
    cpu_period = 100000
  }

  triggers = {
    dir_sha1 = sha1(join("", [for f in fileset(path.module, "build/*") : filesha1(f)]))
  }
}
```
and 
```hcl
resource "docker_container" "workspace" {
  cpu_shares = "500"
```

the limit is not working properly.

![image](https://github.com/coder/coder/assets/4011348/c6823782-62ee-4974-9dd0-e7419b02eed6)

![image](https://github.com/coder/coder/assets/4011348/9c6bb4c3-0660-4cf9-8635-dedd03c06335)

I'm wondering if the Coder community have any solution for this?

**Comment by matifali** [comment, marked as answer]
`cpu_shares` isn't a hard limit it is a relative weightage that only applies if there are not enough resources available. 
For example if you have 2 container running at the same time with `cpu_shares=500` or any equal value then CPU cycles are shared between them equally. 
But let's say if one of them is idle, the other container can then use the full CPU. 
For more details see: https://docs.docker.com/config/containers/resource_constraints/#cpu

It looks like terraform provider only supports [`cpu_shares`](https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container#cpu_shares) and [`cpu_set`](https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container#cpu_set)

**Reply by jizusun** [reply]
Thank you so much!

Do you think the https://github.com/kreuzwerker/terraform-provider-docker/pull/575 would help?

**Reply by matifali** [reply]
Yes, this will be a good addition when merged. But I suggest going with `cpu_shares` as it allows a container to use all of the CPU if no one else is using it.
It is a smarter limit that is only enforced when needed. 

What is your use case? Why do you need a hard limit on CPUs, and `cpu_shares` doesn't serve your purpose?

**Reply by jizusun** [reply]
Here is briefly what I've done to make it work:

```sh
git clone https://github.com/Blackskyliner/terraform-provider-docker.git
cd terraform-provider-docker
git checkout kreuzwerker/terraform-provider-docker/issues/331
make local-install

```
